Verify upstream commits in Makefile

ironcladlou opened the pull request:
Add UPSTREAM/godeps commit validation during the Makefile `verify` task.

Part of https://trello.com/c/noaPjUWG.

ironcladlou commented:
cc @deads2k

deads2k commented on cmd/rebasehelpers/commitchecker/validate.go:
```diff
@@ -0,0 +1,111 @@
+		}
+	}
+	if len(problemCommits) > 0 {
+		label := "No commits were marked UPSTREAM, but the following Godeps changes were detected"
```
This isn't entirely accurate.  I could have an UPSTREAM commit and a normal commit and make godep changes in the normal commit.  This message would then appear.

deads2k replied:
Provide the valid commit format.

ironcladlou replied:
>This isn't entirely accurate. I could have an UPSTREAM commit and a normal commit and make godep changes in the normal commit. This message would then appear.

Isn't that correct then? The commit containing the Godeps changes should have an UPSTREAM commit summary.

deads2k replied:
> Isn't that correct then? The commit containing the Godeps changes should have an UPSTREAM commit summary.

Its correct to produce an error.  The text for that error might be misleading.

commit1= "UPSTREAM: ***"
commit2="foo"

commit2 touched a godep.  I had a commit named UPSTREAM (contradicting the message), my other commit was just wrong.

ironcladlou replied:
OH, I see what you mean now. Thanks!

ironcladlou replied:
Fixed

ironcladlou commented:
@deads2k All cleaned up and integrated with the Makefile, PTAL.

deads2k commented on hack/verify-upstream-commits.sh:
```diff
@@ -0,0 +1,16 @@
+"${OS_ROOT}/hack/build-go.sh" cmd/rebasehelpers/commitchecker
+
+# Find binary
+commitchecker=$( (ls -t _output/local/bin/$(os::build::host_platform)/commitchecker) 2>/dev/null || true | head -1 )
```
yuck, but we're going to want to do it for all of the helpers we build this way.  make a bash function?

ironcladlou replied:
Extracted a function into `common.sh`.

deads2k commented on cmd/rebasehelpers/util/git.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+func (c Commit) DeclaresUpstreamChange() bool {
+	return strings.Contains(strings.ToLower(c.Summary), "upstream")
```
We specifically want "UPSTREAM", no mixed case.

ironcladlou replied:
We validate the case elsewhere: this is just to detect whether a commit summary "looks like an upstream commit," so we want to catch use of the term regardless of case.

liggitt replied:
what would a false positive do here? e.g. "React to upstream changes"

ironcladlou replied:
If you had such a summary with no changes to `Godeps/...` then you'd get dinged with an error. A consequence of the validation strategy overall is that "upstream" becomes a reserved word in summaries, otherwise we can't validate two-ways (undeclared with godeps changes, declared without godeps changes).

deads2k replied:
But couldn't we strictly reserve "starts with UPSTREAM"?

ironcladlou replied:
We could. In the end we're talking about a heuristic, so we have to find some balance. I can contrive summaries that would break the "starts with" approach and some that break the reservation of "upstream" (just consider something like an "uptream" typo).

I'm find with HasPrefix(upstream) if you guys are.

liggitt replied:
+1 for `HasPrefix(..., "UPSTREAM")`

ironcladlou replied:
Switched to a prefix check

deads2k commented on cmd/rebasehelpers/util/git.go:
```diff
@@ -0,0 +1,97 @@
+type File string
+
+func (f File) HasGodepsChanges() bool {
+	return strings.Contains(string(f), "Godeps")
```
Shouldn't this be a `HasPrefix`?

ironcladlou replied:
If we can assume relative paths, yeah. How do you feel about that assumption?

deads2k replied:
> If we can assume relative paths, yeah. How do you feel about that assumption?

not too badly.  The script can always set us up in `OS_ROOT`

ironcladlou replied:
We're talking about the paths represented in the git tree, so HasPrefix should work.

ironcladlou replied:
Switched to a prefix check

deads2k commented on cmd/rebasehelpers/commitchecker/validate.go:
```diff
@@ -0,0 +1,151 @@
+		repos := map[string]struct{}{}
+		for _, file := range commit.Files {
+			if !file.HasGodepsChanges() {
+				continue
```
An upstream commit should never touch a non-godep file, so this should be an error. I don't mind a separate validator.

ironcladlou replied:
Added a new validator for this.

deads2k commented on cmd/rebasehelpers/commitchecker/validate.go:
```diff
@@ -0,0 +1,151 @@
+					break
+				}
+			}
+			// Godeps path struture assumption: Godeps/_workspace/src/repo/name
```
`_workspace/location(code.google, github,etc)/account/repo`

deads2k replied:
Oh, they're all different.  Awesome.

deads2k commented:
Validators I didn't see:
 1. Only updates to k8s.io/kubernetes are allows to skip the account/repo name in the commit message
 1. If a commit specifies an account/repo, it can only modify files in that particular account/repo.

ironcladlou commented:
>Validators I didn't see:
>1. Only updates to k8s.io/kubernetes are allows to skip the account/repo name in the commit message
>2. If a commit specifies an account/repo, it can only modify files in that particular account/repo.

I personally wouldn't mind if we kept things 100% explicit and required the account/repo name in all cases rather than special-casing kube. Makes it easier to implement and analyze, and I don't think requiring the few extra characters is a significant burden.

Thoughts?

deads2k commented:
> I personally wouldn't mind if we kept things 100% explicit and required the account/repo name in all cases rather than special-casing kube. Makes it easier to implement and analyze, and I don't think requiring the few extra characters is a significant burden.
> 
> Thoughts?

We have so many kube patches compared to the rest that I'd rather special case it.

ironcladlou commented:
>We have so many kube patches compared to the rest that I'd rather special case it.

Thinking about this more, I'm still arguing that even kube should be explicit. It's not like people are writing these summaries multiple times a day. It'll be more obvious/consistent when reading, and easier to deal with in code.

deads2k commented:
> Thinking about this more, I'm still arguing that even kube should be explicit. It's not like people are writing these summaries multiple times a day. It'll be more obvious/consistent when reading, and easier to deal with in code.

@liggitt @smarterclayton thoughts?

I'm in favor of special casing kube, so we don't have to add `k8s.io/kubernetes` to almost every UPSTREAM commit.

liggitt commented:
I don't feel strongly either way

smarterclayton commented:
Special case.

On Dec 10, 2015, at 1:17 PM, David Eads <notifications@github.com> wrote:

Thinking about this more, I'm still arguing that even kube should be
explicit. It's not like people are writing these summaries multiple times a
day. It'll be more obvious/consistent when reading, and easier to deal with
in code.

@liggitt <https://github.com/liggitt> @smarterclayton
<https://github.com/smarterclayton> thoughts?

I'm in favor of special casing kube, so we don't have to add
k8s.io/kubernetes to almost every UPSTREAM commit.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/6203#issuecomment-163707544>.

ironcladlou commented:
@deads2k - a few changes:

1. Added a validation to ensure that the declared upstream repo matches the repo(s) changed in the commit
2. Using "k8s.io/kubernetes" as the default repo if none is specified in the summary
3. Refactored to no longer distinguish between the revert and non-revert pattern

deads2k commented on cmd/rebasehelpers/commitchecker/commitchecker.go:
```diff
@@ -0,0 +1,42 @@
+	validators := []func([]util.Commit) error{
+		ValidateUpstreamCommitsWithoutGodepsChanges,
+		ValidateUpstreamCommitModifiesSingleGodepsRepo,
+		ValidateUpstreamCommitSummaries,
```
Looks like you're missing a couple here

ironcladlou replied:
Moved registration of validators into a central location which I now reference here.

deads2k commented on cmd/rebasehelpers/util/git.go:
```diff
@@ -0,0 +1,164 @@
+			break
+		}
+	}
+	// Godeps path struture assumption: Godeps/_workspace/src/repo/name
```
I think you'll have to have different path segments based on the source hosting.  Compare a few to see.  github.com/account/repo is different than k8s.io/heapster

ironcladlou replied:
I believe I've fixed support for this, but the code and regex are now a little more complicated and we have to maintain a map of supported hosts -> expected segment count. Works, but very ugly.

deads2k commented:
Minor comments.  Go ahead a squash.

Also, I'm so glad you didn't do this in bash.

ironcladlou commented:
>Also, I'm so glad you didn't do this in bash.

The idea we could have done this in bash is hilarious in hindsight

deads2k commented:
lgtm, but you need to sort out your verify problems in travis.

deads2k commented:
once you're green, squash and merge at will.

ironcladlou commented:
[merge]
